implement gittuf add-hooks to configure pre-push hook

datosh opened the pull request:
Manage hooks as described in: #158 

Added `UpdateHook` functionality. There doesn't seem to be support for managing hooks in go-git (https://github.com/src-d/go-git/issues?q=is%3Aissue+is%3Aopen+hook) 

~~Unit test currently breaks as the `conf.Core.Worktree` value is not populated when creating a repo with `PlainInit`. Any input appreciated.~~ done

Add following tests / features before merging: 
* make sure hook is executable
* do not overwrite hook script unless force

Further considerations:
* `gittuf rsl record HEAD` should not add a record when executed multiple times. We need a new issue for this.

adityasaky commented on internal/repository/hook.go:
```diff
@@ -0,0 +1,29 @@
+	}
+
+	repoRoot := conf.Core.Worktree
+	hookFile := path.Join(repoRoot, ".git", "hooks", string(hookType))
```
It'd be nice to identify `.git` rather than hardcode it, for cases with detached git-dirs.

datosh replied:
I am not sure what a "detached git-dir" is. Could you provide me an example so I can test against the implementation?

datosh replied:
oh wow, TIL! 

Do you know if there is a better approach than using `r.r.Worktree()` to eventually get to the `hooks` folder? If not then I will add more logic to test if `.git` is dir or file etc...

datosh replied:
Created issue (https://github.com/go-git/go-git/issues/977) and added comment.

adityasaky commented on internal/repository/hook.go:
```diff
@@ -0,0 +1,29 @@
+	repoRoot := conf.Core.Worktree
+	hookFile := path.Join(repoRoot, ".git", "hooks", string(hookType))
+
+	if err := os.WriteFile(hookFile, content, 0o700); err != nil {
```
We should at the very least abort if a pre-push hook already exists. We don't want to nuke other hooks the user may have installed.

spectre10 replied:
As a side note, hooks may not always be scripts, they can be any executable binary. So appending to existing hooks may not work as well.

adityasaky replied:
good point! i think we check and if so, we present the parts of the hook they ought to add to the script themself. Maybe a link to docs someplace.

datosh replied:
Good point. I added the necessary checks to not overwrite existing hooks. 
Do we want to provide a `--force` flag for `gittuf add-hooks` or are we good with the basic functionality for now?

adityasaky replied:
`--force` is fine to add I think, with a warning printed to stderr.

adityasaky commented on internal/cmd/addhooks/prepush.go:
```diff
@@ -0,0 +1,23 @@
+if ! command -v gittuf > /dev/null
+then
+    echo "gittuf could not be found"
+    echo "Download from: https://github.com/gittuf/gittuf"
```
```suggestion
    echo "Download from: https://github.com/gittuf/gittuf/releases/latest"
```

adityasaky commented on internal/cmd/addhooks/prepush.go:
```diff
@@ -0,0 +1,23 @@
+
+echo "Pushing refs/gittuf/reference-state-log with ${remote}"
+# --no-verify disables pre-push hook, to prevent infinite loop.
+git push --no-verify ${remote} refs/gittuf/reference-state-log
```
Interesting, I wonder if `gittuf rsl remote push` bypasses the hook trigger anyway. I'm reasonably certain it should.

datosh replied:
I tried, and somehow `gittuf rsl remote push` does not discover my push credentials. Maybe a bug we should file? I can do more investigation tomorrow.

adityasaky replied:
Very strange, more details would be great for this.

datosh replied:
```
datosh@lenovo-x1:~/code/testing-gittuf/repo$ git push --no-verify origin refs/gittuf/reference-state-log 
Everything up-to-date
datosh@lenovo-x1:~/code/testing-gittuf/repo$ gittuf rsl remote push origin
Error: unable to push RSL
authentication required
Usage:
  gittuf rsl remote push <remote> [flags]

Flags:
  -h, --help   help for push

Global Flags:
      --verbose   enable verbose logging
```

`gittuf` does not seem to provide a lot of details, even with `--verbose` flag.

As I authenticated to this repository using VSCode, and I had similar issues in the past with other tools, I suspect VSCode is doing something non-standard. I will do some investigation, but don't hesitate to provide any pointers :)

adityasaky replied:
The `--verbose` flag doesn't do much yet. We have the logging framework merged as of a week or so ago, but now we need to go in and add Debug statements.

adityasaky replied:
gittuf * push has worked with the standard CLI invocation where I'm authenticated using ssh-agent and so. Nothing jumps out at me atm re the VSCode thing.

datosh replied:
> Interesting, I wonder if `gittuf rsl remote push` bypasses the hook trigger anyway. I'm reasonably certain it should.

Yes, it does. I have changed the hook script to only rely on `gittuf`.

> gittuf * push has worked with the standard CLI invocation where I'm authenticated using ssh-agent and so. Nothing jumps out at me atm re the VSCode thing.

It seems to be related to my particular VSCode setup. I see the same auth issue with other tools.
On the other hand, when I checkout with ssh key everything works fine. I think we can ignore that auth error in this context.

adityasaky reviewed (commented):
Some general comments, I like where this is going!

datosh commented:
I'll fix the DCO before merging to keep the history intact.

adityasaky commented:
Just took another quick look. I think this is largely good to go minus `--force` and a corresponding test.

adityasaky commented on internal/cmd/addhooks/prepush.go:
```diff
@@ -0,0 +1,22 @@
+fi
+
+echo "Pulling RSL from ${remote}."
+gittuf rsl remote pull ${remote}
```
We have a slight issue with branching here where we don't check if rsl update includes changes to the branch currently being pushed. I don't think we can do that strictly within the hook, maybe using a more sophisticated gittuf command.

Since the HEAD branch is not updated but the RSL is, we can't run `verify-ref` after the pull either, because that'll fail simply because the HEAD != expected value in the RSL.

datosh replied:
I see the problem you are describing, but currently we do not run any verification in the script. Is this something we should do?

adityasaky replied:
It's fine for now, I was also thinking aloud here.

datosh commented:
> Just took another quick look. I think this is largely good to go minus `--force` and a corresponding test.

I implemented `--force` and squashed so the DCO check is happy!

adityasaky commented on internal/cmd/addhooks/addhooks.go:
```diff
@@ -0,0 +1,56 @@
+		"force",
+		"f",
+		false,
+		"Overwrite hooks, if they already exist",
```
```suggestion
		"overwrite hooks, if they already exist",
```

adityasaky reviewed (approved):
LGTM, aside a minor consistency nit in the help message for the command.
